### PR TITLE
Week 4 - retrofit on some shit HTTP gods fly on

### PR DESCRIPTION
**Pull request by pelusodan**
Coming in hot fellas!!! 🥵

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -2,6 +2,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="jayp.bge.playboikotlin">
 
+    <uses-permission android:name="android.permission.INTERNET" />
```
Comment (robert-chatterton) [inline app/src/main/AndroidManifest.xml:5]: oh god no not more xml please

File: app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+data class Success(
+    val data: List<User>? = null
+    ) : WebResponse()
+
+data class Failure(
+    val errorMessage: String
+    ) : WebResponse()
+
+object Loading : WebResponse()
```
Comment (robert-chatterton) [inline app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt:62]: !!

File: app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebService.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package jayp.bge.playboikotlin.dan.fragments.webloader
+
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+interface DanWebService {
+
+    @GET("/users/{id}/todos")
```
Comment (robert-chatterton) [inline app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebService.kt:8]: sick

**Review by robert-chatterton — APPROVED** [review]
looks good chief

File: app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.lang.Exception
+import kotlin.properties.Delegates.observable
+import kotlin.reflect.KProperty
+
+const val BASE_URL = "https://jsonplaceholder.typicode.com"
```
Comment (zohebaziz) [inline app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt:15]: classic

File: app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                val data = webService?.getUsers(id)
+                viewState.copy(webResponse = Success(data = data))
+            } catch (e: Exception){
+                viewState.copy(webResponse = Failure(e.message ?: e.toString()))
+            }
+
+        }
+    }
+
+    init {
```
Comment (zohebaziz) [inline app/src/main/java/jayp/bge/playboikotlin/dan/fragments/webloader/DanWebRequestViewModel.kt:38]: bruv

**Review by zohebaziz — APPROVED** [review]

**Pull request merged**